### ISSUE TITLE
/var/ossec/bin/agent-auth disregards giving a hostname w/r/t certificate authentication

### ISSUE THREAD

**Issue by robison** [ossec/ossec-hids#571]
/var/ossec/bin/agent-auth doesn't care whether you give it a hostname or an IP address when it comes to verifying the server certificate presented by ossec-authd.

Specifically, executing ossec-authd on the server:

```
/var/ossec/bin/ossec-authd -p 1515 -v ca.pem -x server.crt -k server.key -d
```

...and executing on the client:

```
/var/ossec/bin/agent-auth -m server.example.com -p 1515 -x client.cer -k client.key -v ca.pem -d
```

...will result in the following logs on the server:

```
2015/04/07 01:16:00 ossec-authd: DEBUG: Starting ...
2015/04/07 01:16:00 ossec-authd: INFO: Started (pid: 9922).
2015/04/07 01:16:00 ossec-authd: DEBUG: Peer verification requested.
2015/04/07 01:16:00 ossec-authd: DEBUG: Returning CTX for server.
2015/04/07 01:16:00 ossec-authd: DEBUG: Going into listening mode.
2015/04/07 01:16:18 ossec-authd: INFO: New connection from 10.123.128.129
2015/04/07 01:16:18 ossec-authd: ERROR: SSL Error (0)
2015/04/07 01:16:18 ossec-authd: DEBUG: Process 9927 exited
```

...and client:

```
2015/04/07 01:17:26 ossec-authd: DEBUG: Starting ...
2015/04/07 01:17:26 ossec-authd: INFO: Started (pid: 1042).
2015/04/07 01:17:26 ossec-authd: DEBUG: Peer verification requested.
2015/04/07 01:17:26 ossec-authd: DEBUG: Returning CTX for client.
INFO: Connected to 1.2.3.4:1515
INFO: Verifing manager's certificate
2015/04/07 01:17:26 ossec-authd: DEBUG: Checking certificate's subject alternative names.
2015/04/07 01:17:26 ossec-authd: DEBUG: No matching subject alternative names found. Checking common name.
2015/04/07 01:17:26 ossec-authd: DEBUG: Unable to verify server certificate.
user@client:~$
```

I've validated that ossec-authd authenticates itself correctly via TLS with:

```
openssl s_client -connect server.example.com:1515 -servername server.example.com -cert client.cer -key client.key -CAfile ca.pem
```

which returns a valid, strong TLS connection:

```
New, TLSv1/SSLv3, Cipher is AES256-GCM-SHA384
Server public key is 4096 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : AES256-GCM-SHA384
    Session-ID: 7A8B0C6E3A1E1BE4AADABB1D2513A560B2A284A63FF4918AC3C0130803C274A9
    Session-ID-ctx:
    Master-Key: DEF7301E1E09880B2C32B63193DAA8499B0EB4F64F770F2925FCC0BE954D5755848EE3C71457DBC816E50B1D380352C2
    Key-Arg   : None
    Krb5 Principal: None
    PSK identity: None
    PSK identity hint: None
    TLS session ticket lifetime hint: 7200 (seconds)
```

TL;DR: agent-auth doesn't keep track of the server's hostname, instead failing the authentication checks if the server doesn't have the IP address in a subject alternative name.

**Comment by Namsep** [comment]
Sorry that I did not check, but out of curiosity does this fix respect IPv6 addresses? As I'm testing the IPv6 support branch I was wondering about this.

**Comment by robison** [comment]
I've not verified that, but based on the logic involved, I believe that it should work splendidly.

I'm redoing my changes to clean up the utter mess I've made of my own repos; I'll resubmit a PR this week.

**Comment by mweigel** [comment]
Hi, gald to see that someone else found this functionality useful but sorry to see that it's not working for you. I've pulled the latest code from master this afternoon and have been trying to replicate your
issue without success. I'm testing on Ubuntu 14.04.2 LTS with OpenSSL 1.0.1f.


./agent-auth -d -m server.example.com -v /var/ossec/ssl/ca.cert.pem

2015/04/25 15:33:16 ossec-authd: DEBUG: Starting ...
2015/04/25 15:33:16 ossec-authd: INFO: Started (pid: 8559).
2015/04/25 15:33:16 ossec-authd: DEBUG: Peer verification requested.
2015/04/25 15:33:16 ossec-authd: DEBUG: Returning CTX for client.
INFO: Connected to 192.168.122.210:1515
INFO: Verifing manager's certificate
2015/04/25 15:33:16 ossec-authd: DEBUG: Checking certificate's subject alternative names.
2015/04/25 15:33:16 ossec-authd: DEBUG: No matching subject alternative names found. Checking common name.
INFO: Using agent name as: oclient
INFO: Send request to manager. Waiting for reply.
INFO: Received response with agent key
INFO: Valid key created. Finished.
INFO: Connection closed.

This works for me regardless of whether the manager's name is in the common name or in a subject alt name field. One issue that comes to mind is that at the moment the code is pretty strict (possibly too strict) about what it will accept as a valid DNS name. Each label must start with an alphabetic character, end with an alphabetic character or digit and can only contain alphabetic characters, digits or hyphens. This could be relaxed pretty easily.

Would you be able to post a printout of the managers's certificate as displayed by openssl x509? If you're not comfortable doing that could you add some extra debugging in the "label_valid" function and give it another try. I'd like to help get to the bottom of this one, cheers.

**Comment by mweigel** [comment]
Also it would be great if someone who is using IPv6 could test this. I've done a fair amount of testing with IPv4 addresses and don't know of any issues there at the moment.

**Comment by Namsep** [comment]
I tried to test it, runned into "ERROR: Not compiled. Missing OpenSSL support."  Also a bit short on time, but running largely IPv6 only I should make time for this somewhere this week.

**Comment by mweigel** [comment]
Make sure the OpenSSL development packages for your OS are installed before building. That should fix the error, cheers.
